state/protocol/badger/mutator.go

```diff
@@ -981,6 +996,73 @@ func (m *FollowerState) epochStatus(block *flow.Header, epochFallbackTriggered b
 
 }
 
+// versionBeaconOnBlockFinalized extracts the VersionBeacons from the parent block's
+// Seals and returns it.
+// This could return multiple VersionBeacons if the parent block contains multiple Seals.
+// The version beacons will be returned in the ascending height order of the seals.
+// Technically only the last seal is relevant.
+func (m *FollowerState) versionBeaconOnBlockFinalized(
+	header *flow.Header,
+) ([]*flow.SealedVersionBeacon, error) {
+	var versionBeacons []*flow.SealedVersionBeacon
+
+	parent, err := m.blocks.ByID(header.ParentID)
```

Comment:
A few comments:
1. Too many db calls
We would like to find every single `VersionBeacon` service event. In order to do that we scan through each sealed result. But we are doing a lot of db calls for that. We first find the block and throw the entire payload away, but only keep the `Payload.Seals` field, and then calling `protocol.OrderedSeals` to find seals, which includes quite a few DB calls to fetch headers, and then for each seals, we lookup each result for it.
Is there way to reduce the db calls? Does the caller have some data queried already? 

2. There are other service events, which will also to go through each sealed results, instead of go through so many db calls for each different service event, can we split it into two steps:
Step1, find newly sealed results
Step2, handle the service events from each sealed result, in the future, if we need to handle another service event, we just add another filter in this step.

Reply:
Idea, since VersionBeacon event is very rare, we are most likely making a lot of queries here, and end up finding no event.

What about building a separate index for the version beacon events, for instance. whenever, we store a result, we also create an index for the beacon event by block ID, that why we only spend 1 db call when there is no beacon event for the certain block.

Reply:
That is a good point. I it its ok with you can I do optimizations as a separate PR? I checked the rest of the code here, and there is already a lot of database access, but the gcp profiler says we don't actually spend a lot of time here, so I don't think this will make a drastic difference.

what I plan to look into is what you mentioned.: creating another index for finalized version beacons, but also looking into unifying places in `Finalize` that have the pattern of taking the finalized seals and getting results for them y ID. There are 3 such places.